File: core/pipeline/startup_diagnostics.py
"""Startup media diagnostics for automatic pipeline selection."""
from __future__ import annotations

import copy
import json
import os
import subprocess
from datetime import datetime
from typing import Any

from core.media_info import probe_media
from core.native_swift_startup_diagnostics import (
    attach_expected_processing_time_via_swift,
    build_startup_diagnostic_via_swift,
    format_startup_diagnostic_log_via_swift,
)
from core.platform_compat import ffprobe_binary, hidden_subprocess_kwargs
from core.project.project_io import read_project_file, write_project_file

STARTUP_DIAGNOSTIC_SCHEMA = "ai_subtitle_studio.startup_diagnostic.v1"


def _to_float(value: Any, default: float = 0.0) -> float:
    try:
        parsed = float(value)
        if parsed == parsed:
            return parsed
    except Exception:
        pass
    return default


def _to_int(value: Any, default: int = 0) -> int:
    try:
        return int(float(value))
    except Exception:
        return default


def _duration_label(seconds: float) -> str:
    seconds = _to_float(seconds, 0.0)
    if seconds <= 0:
        return "-"
    total = int(round(seconds))
    minutes, sec = divmod(total, 60)
    hours, minutes = divmod(minutes, 60)
    if hours:
        return f"{hours:02d}:{minutes:02d}:{sec:02d}"
    return f"{minutes:02d}:{sec:02d}"


def _processing_label(seconds: float) -> str:
    seconds = _to_float(seconds, 0.0)
    if seconds <= 0:
        return "예상불가"
    total = int(round(seconds))
    minutes, sec = divmod(total, 60)
    hours, minutes = divmod(minutes, 60)
    if hours:
        return f"{hours}시간 {minutes}분 {sec}초"
    if minutes:
        return f"{minutes}분 {sec}초"
    return f"{sec}초"


def _boundary_time(row: Any) -> float:
    if isinstance(row, dict):
        for key in ("timeline_sec", "time", "start", "sec"):
            if key in row:
                return _to_float(row.get(key), 0.0)
        return 0.0
    return _to_float(row, 0.0)


def _first_audio_stream(payload: dict[str, Any]) -> dict[str, Any] | None:
    for stream in list(payload.get("streams", []) or []):
        if isinstance(stream, dict) and stream.get("codec_type") == "audio":
            return stream
    return None


def probe_audio_stream_info(media_path: str, *, timeout_sec: float = 5.0) -> dict[str, Any]:
    """Return lightweight audio metadata used by the start diagnostic."""
    result: dict[str, Any] = {
        "has_audio": False,
        "codec": "",
        "sample_rate": 0,
        "channels": 0,
        "bit_rate": 0,
        "duration_sec": 0.0,
    }
    try:
        cmd = [
            ffprobe_binary(),
            "-v",
            "error",
            "-show_entries",
            "format=duration,bit_rate:stream=index,codec_type,codec_name,sample_rate,channels,bit_rate,duration",
            "-of",
            "json",
            media_path,
        ]
        proc = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=timeout_sec,
            **hidden_subprocess_kwargs(),
        )
        payload = json.loads(proc.stdout or "{}")
        stream = _first_audio_stream(payload)
        if not stream:
            return result
        fmt = payload.get("format", {}) if isinstance(payload.get("format"), dict) else {}
        result.update(
            {
                "has_audio": True,
                "codec": str(stream.get("codec_name", "") or ""),
                "sample_rate": _to_int(stream.get("sample_rate"), 0),
                "channels": _to_int(stream.get("channels"), 0),
                "bit_rate": _to_int(stream.get("bit_rate"), _to_int(fmt.get("bit_rate"), 0)),
                "duration_sec": _to_float(stream.get("duration"), _to_float(fmt.get("duration"), 0.0)),
            }
        )
    except Exception:
        pass
    return result


def _audio_quality(audio_info: dict[str, Any]) -> dict[str, Any]:
    if not bool(audio_info.get("has_audio")):
        return {
            "score": 0,
            "label": "red",
            "summary": "오디오 없음",
            "noise_estimate": "high",
            "noise_label": "높음",
            "source": "metadata_heuristic",
            "reasons": ["audio_missing"],
        }

    score = 100
    reasons: list[str] = []
    sample_rate = _to_int(audio_info.get("sample_rate"), 0)
    bit_rate = _to_int(audio_info.get("bit_rate"), 0)
    channels = _to_int(audio_info.get("channels"), 0)

    if sample_rate and sample_rate < 16000:
        score -= 35
        reasons.append("low_sample_rate")
    elif sample_rate and sample_rate < 32000:
        score -= 15
        reasons.append("medium_sample_rate")
    elif sample_rate <= 0:
        score -= 20
        reasons.append("unknown_sample_rate")

    if bit_rate and bit_rate < 48000:
        score -= 30
        reasons.append("low_bit_rate")
    elif bit_rate and bit_rate < 96000:
        score -= 12
        reasons.append("medium_bit_rate")
    elif bit_rate <= 0:
        score -= 8
        reasons.append("unknown_bit_rate")

    if channels <= 0:
        score -= 8
        reasons.append("unknown_channels")

    score = max(0, min(100, score))
    if score >= 75:
        label, summary, noise, noise_label = "green", "양호", "low", "낮음"
    elif score >= 45:
        label, summary, noise, noise_label = "yellow", "주의", "medium", "중간"
    else:
        label, summary, noise, noise_label = "red", "복구 필요", "high", "높음"

    return {
        "score": score,
        "label": label,
        "summary": summary,
        "noise_estimate": noise,
        "noise_label": noise_label,
        "source": "metadata_heuristic",
        "reasons": reasons,
    }


def _cut_density_profile(
    cut_boundaries: list[Any] | None,
    provisional_boundaries: list[Any] | None,
    duration_sec: float,
) -> dict[str, Any]:
    cuts = [_boundary_time(row) for row in list(cut_boundaries or [])]
    cuts = [sec for sec in cuts if sec > 0]
    provisional = [_boundary_time(row) for row in list(provisional_boundaries or [])]
    provisional = [sec for sec in provisional if sec > 0]
    minutes = max(_to_float(duration_sec, 0.0) / 60.0, 0.0)
    per_minute = (len(cuts) / minutes) if minutes > 0 else 0.0
    if per_minute >= 4.0:
        level = "high"
        label = "높음"
    elif per_minute >= 1.0:
        level = "medium"
        label = "중간"
    else:
        level = "low"
        label = "낮음"
    return {
        "verified_count": len(cuts),
        "provisional_count": len(provisional),
        "per_minute": round(per_minute, 3),
        "level": level,
        "label": label,
    }


def _speaker_hint(settings: dict[str, Any] | None, speaker_count_hint: int | None) -> dict[str, Any]:
    if speaker_count_hint is not None:
        return {"count": max(1, _to_int(speaker_count_hint, 1)), "source": "runtime"}
    settings = settings if isinstance(settings, dict) else {}
    return {"count": max(1, _to_int(settings.get("max_speakers"), 1)), "source": "settings"}


def _recommend_pipeline(
    *,
    duration_sec: float,
    fps: float,
    audio_quality: dict[str, Any],
    cut_density: dict[str, Any],
    speaker_count: int,
) -> dict[str, Any]:
    if duration_sec <= 0:
        return {
            "mode": "recovery",
            "label": "복구 모드",
            "score": 100,
            "reasons": ["duration_unknown"],
        }
    if int(audio_quality.get("score", 0) or 0) < 35:
        return {
            "mode": "recovery",
            "label": "복구 모드",
            "score": 95,
            "reasons": ["audio_quality_low"],
        }

    score = 0
    reasons: list[str] = []
    if duration_sec >= 20 * 60:
        score += 2
        reasons.append("long_video")
    elif duration_sec <= 5 * 60:
        score -= 1
        reasons.append("short_video")
    if fps >= 50:
        score += 1
        reasons.append("high_fps")
    if str(cut_density.get("level")) == "high":
        score += 2
        reasons.append("dense_cuts")
    elif str(cut_density.get("level")) == "low" and duration_sec <= 8 * 60:
        score -= 1
        reasons.append("simple_cut_structure")
    if speaker_count >= 2:
        score += 1
        reasons.append("multi_speaker")
    if int(audio_quality.get("score", 0) or 0) < 70:
        score += 1
        reasons.append("audio_attention_needed")

    if score <= -2:
        mode, label = "fast", "빠른 모드"
    elif score <= 1:
        mode, label = "balanced", "균형 모드"
    else:
        mode, label = "precise", "정밀 모드"
    return {"mode": mode, "label": label, "score": score, "reasons": reasons}


def build_startup_diagnostic(
    media_path: str,
    *,
    settings: dict[str, Any] | None = None,
    cut_boundaries: list[Any] | None = None,
    provisional_cut_boundaries: list[Any] | None = None,
    expected_time_sec: float | None = None,
    speaker_count_hint: int | None = None,
) -> dict[str, Any]:
    """Build a start-of-run diagnostic profile for a media file."""
    media = probe_media(media_path)
    duration_sec = _to_float(media.get("duration"), 0.0)
    fps = _to_float(media.get("fps"), 0.0)
    audio_info = probe_audio_stream_info(media_path)
    native = build_startup_diagnostic_via_swift(
        media_path,
        media={
            "duration_sec": duration_sec,
            "fps": fps,
            "width": _to_int(media.get("width"), 0),
            "height": _to_int(media.get("height"), 0),
            "info_txt": str(media.get("info_txt", "") or ""),
        },
        audio=audio_info,
        settings=settings,
        cut_boundaries=cut_boundaries,
        provisional_cut_boundaries=provisional_cut_boundaries,
        expected_time_sec=expected_time_sec,
        speaker_count_hint=speaker_count_hint,
    )
    if isinstance(native, dict) and native.get("schema") == STARTUP_DIAGNOSTIC_SCHEMA:
        return native
    audio_quality = _audio_quality(audio_info)
    cut_density = _cut_density_profile(cut_boundaries, provisional_cut_boundaries, duration_sec)
    speakers = _speaker_hint(settings, speaker_count_hint)
    recommendation = _recommend_pipeline(
        duration_sec=duration_sec,
        fps=fps,
        audio_quality=audio_quality,
        cut_density=cut_density,
        speaker_count=int(speakers.get("count", 1) or 1),
    )
    expected = _to_float(expected_time_sec, 0.0)

    return {
        "schema": STARTUP_DIAGNOSTIC_SCHEMA,
        "created_at": datetime.now().isoformat(timespec="seconds"),
        "media_path": media_path,
        "media_name": os.path.basename(media_path),
        "media": {
            "duration_sec": round(duration_sec, 3),
            "duration_label": _duration_label(duration_sec),
            "fps": round(fps, 3),
            "width": _to_int(media.get("width"), 0),
            "height": _to_int(media.get("height"), 0),
            "info_txt": str(media.get("info_txt", "") or ""),
        },
        "audio": {**audio_info, "quality": audio_quality},
        "speakers": speakers,
        "cut_density": cut_density,
        "estimated_processing_sec": round(expected, 3) if expected > 0 else 0.0,
        "estimated_processing_label": _processing_label(expected),
        "estimated_processing_source": "history" if expected > 0 else "unknown",
        "recommended_pipeline": recommendation,
    }


def attach_expected_processing_time(
    diagnostic: dict[str, Any],
    expected_time_sec: float,
    *,
    source: str = "history",
) -> dict[str, Any]:
    """Return a diagnostic copy with ETA fields attached."""
    native = attach_expected_processing_time_via_swift(
        diagnostic if isinstance(diagnostic, dict) else {},
        expected_time_sec,
        source=source,
    )
    if isinstance(native, dict):
        return native
    updated = copy.deepcopy(diagnostic if isinstance(diagnostic, dict) else {})
    expected = _to_float(expected_time_sec, 0.0)
    updated["estimated_processing_sec"] = round(expected, 3) if expected > 0 else 0.0
    updated["estimated_processing_label"] = _processing_label(expected)
    updated["estimated_processing_source"] = str(source or "history") if expected > 0 else "unknown"
    return updated


def _reason_label(reason: str) -> str:
    labels = {
        "duration_unknown": "길이 확인 필요",
        "audio_quality_low": "오디오 품질 낮음",
        "long_video": "긴 영상",
        "short_video": "짧은 영상",
        "high_fps": "고FPS",
        "dense_cuts": "컷 많음",
        "simple_cut_structure": "단순 컷",
        "multi_speaker": "복수 화자",
        "audio_attention_needed": "오디오 주의",
    }
    return labels.get(str(reason), str(reason))


def format_startup_diagnostic_log(diagnostic: dict[str, Any]) -> list[str]:
    """Format concise Korean log lines for the terminal/status panel."""
    native = format_startup_diagnostic_log_via_swift(diagnostic if isinstance(diagnostic, dict) else {})
    if isinstance(native, list):
        return native
    if not isinstance(diagnostic, dict):
        return []
    media = diagnostic.get("media", {}) or {}
    audio = diagnostic.get("audio", {}) or {}
    quality = audio.get("quality", {}) or {}
    cut_density = diagnostic.get("cut_density", {}) or {}
    recommendation = diagnostic.get("recommended_pipeline", {}) or {}
    reasons = ", ".join(_reason_label(row) for row in list(recommendation.get("reasons", []) or [])[:4])
    if not reasons:
        reasons = "기본 안정값"
    sample_rate = _to_int(audio.get("sample_rate"), 0)
    channels = _to_int(audio.get("channels"), 0)
    sample_label = f"{sample_rate / 1000:.1f}kHz" if sample_rate else "샘플레이트 미상"
    channel_label = f"{channels}ch" if channels else "채널 미상"
    return [
        (
            "  🩺 [시작 진단] "
            f"{diagnostic.get('media_name', '')} · {media.get('duration_label', '-')} · "
            f"{media.get('fps', 0):.2f}fps · {media.get('width', 0)}x{media.get('height', 0)}"
        ),
        (
            "  🩺 [시작 진단] "
            f"오디오 {quality.get('summary', '미상')} · 노이즈 추정 {quality.get('noise_label', '미상')} · "
            f"{sample_label} · {channel_label}"
        ),
        (
            "  🩺 [시작 진단] "
            f"컷 밀도 {cut_density.get('label', '미상')} · 정식 {cut_density.get('verified_count', 0)}개 · "
            f"임시 {cut_density.get('provisional_count', 0)}개 · {cut_density.get('per_minute', 0.0):.2f}/분"
        ),
        (
            "  🩺 [시작 진단] "
            f"추천 {recommendation.get('label', '균형 모드')} · 예상 {diagnostic.get('estimated_processing_label', '예상불가')} · "
            f"근거: {reasons}"
        ),
    ]


def persist_startup_diagnostic(project_path: str, diagnostic: dict[str, Any]) -> bool:
    """Persist startup diagnostics into both project analysis locations."""
    if not project_path or not os.path.exists(project_path) or not isinstance(diagnostic, dict):
        return False
    try:
        project = read_project_file(project_path)
        analysis = project.setdefault("analysis", {})
        analysis["startup_diagnostic_schema"] = STARTUP_DIAGNOSTIC_SCHEMA
        analysis["startup_diagnostic"] = diagnostic

        editor_state = project.setdefault("editor_state", {})
        editor_analysis = editor_state.setdefault("analysis", {})
        editor_analysis["startup_diagnostic_schema"] = STARTUP_DIAGNOSTIC_SCHEMA
        editor_analysis["startup_diagnostic"] = copy.deepcopy(diagnostic)
        write_project_file(project_path, project)
        return True
    except Exception:
        return False

File: native/macos/AIStudioNative/Sources/AIStudioCore/StartupDiagnostics.swift
import Foundation

public enum StartupDiagnosticsNative {
    public static func build(payload: [String: Any]) -> [String: Any] {
        let mediaPath = stringValue(payload["media_path"])
        let mediaName = stringValue(payload["media_name"], default: URL(fileURLWithPath: mediaPath).lastPathComponent)
        let media = dictValue(payload["media"])
        let audio = dictValue(payload["audio"])
        let settings = dictValue(payload["settings"])
        let cutBoundaries = arrayValue(payload["cut_boundaries"])
        let provisionalBoundaries = arrayValue(payload["provisional_cut_boundaries"])
        let expectedTimeSec = doubleValue(payload["expected_time_sec"])
        let speakerCountHint = payload["speaker_count_hint"]

        let durationSec = doubleValue(media["duration_sec"], default: doubleValue(media["duration"]))
        let fps = doubleValue(media["fps"])
        let audioQuality = audioQualityProfile(audio)
        let cutDensity = cutDensityProfile(
            cutBoundaries: cutBoundaries,
            provisionalBoundaries: provisionalBoundaries,
            durationSec: durationSec
        )
        let speakers = speakerHint(settings: settings, hint: speakerCountHint)
        let recommendation = recommendPipeline(
            durationSec: durationSec,
            fps: fps,
            audioQuality: audioQuality,
            cutDensity: cutDensity,
            speakerCount: intValue(speakers["count"], default: 1)
        )
        let expected = max(0.0, expectedTimeSec)

        return [
            "schema": "ai_subtitle_studio.startup_diagnostic.v1",
            "created_at": createdAtString(),
            "media_path": mediaPath,
            "media_name": mediaName,
            "media": [
                "duration_sec": roundTo(durationSec, digits: 3),
                "duration_label": durationLabel(durationSec),
                "fps": roundTo(fps, digits: 3),
                "width": intValue(media["width"]),
                "height": intValue(media["height"]),
                "info_txt": stringValue(media["info_txt"]),
            ],
            "audio": merge(audio, with: ["quality": audioQuality]),
            "speakers": speakers,
            "cut_density": cutDensity,
            "estimated_processing_sec": expected > 0.0 ? roundTo(expected, digits: 3) : 0.0,
            "estimated_processing_label": processingLabel(expected),
            "estimated_processing_source": expected > 0.0 ? "history" : "unknown",
            "recommended_pipeline": recommendation,
        ]
    }

    public static func attachExpected(payload: [String: Any]) -> [String: Any] {
        var diagnostic = dictValue(payload["diagnostic"])
        let expected = max(0.0, doubleValue(payload["expected_time_sec"]))
        let source = stringValue(payload["source"], default: "history")
        diagnostic["estimated_processing_sec"] = expected > 0.0 ? roundTo(expected, digits: 3) : 0.0
        diagnostic["estimated_processing_label"] = processingLabel(expected)
        diagnostic["estimated_processing_source"] = expected > 0.0 ? source : "unknown"
        return diagnostic
    }

    public static func formatLog(payload: [String: Any]) -> [String: Any] {
        let diagnostic = dictValue(payload["diagnostic"])
        let media = dictValue(diagnostic["media"])
        let audio = dictValue(diagnostic["audio"])
        let quality = dictValue(audio["quality"])
        let cutDensity = dictValue(diagnostic["cut_density"])
        let recommendation = dictValue(diagnostic["recommended_pipeline"])
        let reasons = arrayValue(recommendation["reasons"])
        let reasonText = reasons
            .prefix(4)
            .map { reasonLabel(stringValue($0)) }
            .filter { !$0.isEmpty }
            .joined(separator: ", ")
        let joinedReasons = reasonText.isEmpty ? "기본 안정값" : reasonText

        let sampleRate = intValue(audio["sample_rate"])
        let channels = intValue(audio["channels"])
        let sampleLabel = sampleRate > 0 ? String(format: "%.1fkHz", Double(sampleRate) / 1000.0) : "샘플레이트 미상"
        let channelLabel = channels > 0 ? "\(channels)ch" : "채널 미상"

        let perMinuteLabel = String(format: "%.2f", doubleValue(cutDensity["per_minute"]))
        let lines = [
            "  🩺 [시작 진단] \(stringValue(diagnostic["media_name"])) · \(stringValue(media["duration_label"], default: "-")) · \(String(format: "%.2f", doubleValue(media["fps"])))fps · \(intValue(media["width"]))x\(intValue(media["height"]))",
            "  🩺 [시작 진단] 오디오 \(stringValue(quality["summary"], default: "미상")) · 노이즈 추정 \(stringValue(quality["noise_label"], default: "미상")) · \(sampleLabel) · \(channelLabel)",
            "  🩺 [시작 진단] 컷 밀도 \(stringValue(cutDensity["label"], default: "미상")) · 정식 \(intValue(cutDensity["verified_count"]))개 · 임시 \(intValue(cutDensity["provisional_count"]))개 · \(perMinuteLabel)/분",
            "  🩺 [시작 진단] 추천 \(stringValue(recommendation["label"], default: "균형 모드")) · 예상 \(stringValue(diagnostic["estimated_processing_label"], default: "예상불가")) · 근거: \(joinedReasons)",
        ]
        return ["lines": lines]
    }

    private static func createdAtString() -> String {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
        return formatter.string(from: Date())
    }

    private static func durationLabel(_ seconds: Double) -> String {
        guard seconds > 0 else { return "-" }
        let total = Int(seconds.rounded())
        let minutes = total / 60
        let sec = total % 60
        let hours = minutes / 60
        let remMinutes = minutes % 60
        if hours > 0 {
            return String(format: "%02d:%02d:%02d", hours, remMinutes, sec)
        }
        return String(format: "%02d:%02d", remMinutes, sec)
    }

    private static func processingLabel(_ seconds: Double) -> String {
        guard seconds > 0 else { return "예상불가" }
        let total = Int(seconds.rounded())
        let minutes = total / 60
        let sec = total % 60
        let hours = minutes / 60
        let remMinutes = minutes % 60
        if hours > 0 {
            return "\(hours)시간 \(remMinutes)분 \(sec)초"
        }
        if remMinutes > 0 {
            return "\(remMinutes)분 \(sec)초"
        }
        return "\(sec)초"
    }

    private static func boundaryTime(_ item: Any) -> Double {
        if let dict = item as? [String: Any] {
            for key in ["timeline_sec", "time", "start", "sec"] {
                if dict[key] != nil {
                    return doubleValue(dict[key])
                }
            }
            return 0.0
        }
        return doubleValue(item)
    }

    private static func audioQualityProfile(_ audioInfo: [String: Any]) -> [String: Any] {
        if !boolValue(audioInfo["has_audio"]) {
            return [
                "score": 0,
                "label": "red",
                "summary": "오디오 없음",
                "noise_estimate": "high",
                "noise_label": "높음",
                "source": "metadata_heuristic",
                "reasons": ["audio_missing"],
            ]
        }

        var score = 100
        var reasons: [String] = []
        let sampleRate = intValue(audioInfo["sample_rate"])
        let bitRate = intValue(audioInfo["bit_rate"])
        let channels = intValue(audioInfo["channels"])

        if sampleRate > 0 && sampleRate < 16_000 {
            score -= 35
            reasons.append("low_sample_rate")
        } else if sampleRate > 0 && sampleRate < 32_000 {
            score -= 15
            reasons.append("medium_sample_rate")
        } else if sampleRate <= 0 {
            score -= 20
            reasons.append("unknown_sample_rate")
        }

        if bitRate > 0 && bitRate < 48_000 {
            score -= 30
            reasons.append("low_bit_rate")
        } else if bitRate > 0 && bitRate < 96_000 {
            score -= 12
            reasons.append("medium_bit_rate")
        } else if bitRate <= 0 {
            score -= 8
            reasons.append("unknown_bit_rate")
        }

        if channels <= 0 {
            score -= 8
            reasons.append("unknown_channels")
        }

        score = max(0, min(100, score))
        let label: String
        let summary: String
        let noise: String
        let noiseLabel: String
        if score >= 75 {
            label = "green"
            summary = "양호"
            noise = "low"
            noiseLabel = "낮음"
        } else if score >= 45 {
            label = "yellow"
            summary = "주의"
            noise = "medium"
            noiseLabel = "중간"
        } else {
            label = "red"
            summary = "복구 필요"
            noise = "high"
            noiseLabel = "높음"
        }

        return [
            "score": score,
            "label": label,
            "summary": summary,
            "noise_estimate": noise,
            "noise_label": noiseLabel,
            "source": "metadata_heuristic",
            "reasons": reasons,
        ]
    }

    private static func cutDensityProfile(
        cutBoundaries: [Any],
        provisionalBoundaries: [Any],
        durationSec: Double
    ) -> [String: Any] {
        let cuts = cutBoundaries.map(boundaryTime).filter { $0 > 0.0 }
        let provisional = provisionalBoundaries.map(boundaryTime).filter { $0 > 0.0 }
        let minutes = max(durationSec / 60.0, 0.0)
        let perMinute = minutes > 0.0 ? Double(cuts.count) / minutes : 0.0
        let level: String
        let label: String
        if perMinute >= 4.0 {
            level = "high"
            label = "높음"
        } else if perMinute >= 1.0 {
            level = "medium"
            label = "중간"
        } else {
            level = "low"
            label = "낮음"
        }
        return [
            "verified_count": cuts.count,
            "provisional_count": provisional.count,
            "per_minute": roundTo(perMinute, digits: 3),
            "level": level,
            "label": label,
        ]
    }

    private static func speakerHint(settings: [String: Any], hint: Any?) -> [String: Any] {
        if let hint, !(hint is NSNull) {
            return [
                "count": max(1, intValue(hint, default: 1)),
                "source": "runtime",
            ]
        }
        return [
            "count": max(1, intValue(settings["max_speakers"], default: 1)),
            "source": "settings",
        ]
    }

    private static func recommendPipeline(
        durationSec: Double,
        fps: Double,
        audioQuality: [String: Any],
        cutDensity: [String: Any],
        speakerCount: Int
    ) -> [String: Any] {
        if durationSec <= 0.0 {
            return [
                "mode": "recovery",
                "label": "복구 모드",
                "score": 100,
                "reasons": ["duration_unknown"],
            ]
        }
        if intValue(audioQuality["score"]) < 35 {
            return [
                "mode": "recovery",
                "label": "복구 모드",
                "score": 95,
                "reasons": ["audio_quality_low"],
            ]
        }

        var score = 0
        var reasons: [String] = []
        if durationSec >= 20.0 * 60.0 {
            score += 2
            reasons.append("long_video")
        } else if durationSec <= 5.0 * 60.0 {
            score -= 1
            reasons.append("short_video")
        }
        if fps >= 50.0 {
            score += 1
            reasons.append("high_fps")
        }
        let cutLevel = stringValue(cutDensity["level"])
        if cutLevel == "high" {
            score += 2
            reasons.append("dense_cuts")
        } else if cutLevel == "low" && durationSec <= 8.0 * 60.0 {
            score -= 1
            reasons.append("simple_cut_structure")
        }
        if speakerCount >= 2 {
            score += 1
            reasons.append("multi_speaker")
        }
        if intValue(audioQuality["score"]) < 70 {
            score += 1
            reasons.append("audio_attention_needed")
        }

        let mode: String
        let label: String
        if score <= -2 {
            mode = "fast"
            label = "빠른 모드"
        } else if score <= 1 {
            mode = "balanced"
            label = "균형 모드"
        } else {
            mode = "precise"
            label = "정밀 모드"
        }
        return [
            "mode": mode,
            "label": label,
            "score": score,
            "reasons": reasons,
        ]
    }

    private static func reasonLabel(_ reason: String) -> String {
        switch reason {
        case "duration_unknown":
            return "길이 확인 필요"
        case "audio_quality_low":
            return "오디오 품질 낮음"
        case "long_video":
            return "긴 영상"
        case "short_video":
            return "짧은 영상"
        case "high_fps":
            return "고FPS"
        case "dense_cuts":
            return "컷 많음"
        case "simple_cut_structure":
            return "단순 컷"
        case "multi_speaker":
            return "복수 화자"
        case "audio_attention_needed":
            return "오디오 주의"
        default:
            return reason
        }
    }

    private static func roundTo(_ value: Double, digits: Int) -> Double {
        guard digits >= 0 else { return value }
        let scale = pow(10.0, Double(digits))
        return (value * scale).rounded() / scale
    }

    private static func merge(_ base: [String: Any], with extra: [String: Any]) -> [String: Any] {
        var out = base
        for (key, value) in extra {
            out[key] = value
        }
        return out
    }

    private static func dictValue(_ value: Any?) -> [String: Any] {
        value as? [String: Any] ?? [:]
    }

    private static func arrayValue(_ value: Any?) -> [Any] {
        value as? [Any] ?? []
    }

    private static func stringValue(_ value: Any?, default defaultValue: String = "") -> String {
        guard let value else { return defaultValue }
        let text = String(describing: value)
        return text.isEmpty ? defaultValue : text
    }

    private static func boolValue(_ value: Any?) -> Bool {
        switch value {
        case let value as Bool:
            return value
        case let value as NSNumber:
            return value.boolValue
        case let value as String:
            return ["1", "true", "yes", "on", "사용", "켜짐"].contains(value.lowercased())
        default:
            return false
        }
    }

    private static func doubleValue(_ value: Any?, default defaultValue: Double = 0.0) -> Double {
        switch value {
        case let value as Double:
            return value.isFinite ? value : defaultValue
        case let value as NSNumber:
            let parsed = value.doubleValue
            return parsed.isFinite ? parsed : defaultValue
        case let value as String:
            let parsed = Double(value) ?? defaultValue
            return parsed.isFinite ? parsed : defaultValue
        default:
            return defaultValue
        }
    }

    private static func intValue(_ value: Any?, default defaultValue: Int = 0) -> Int {
        switch value {
        case let value as Int:
            return value
        case let value as NSNumber:
            return value.intValue
        case let value as String:
            return Int(Double(value) ?? Double(defaultValue))
        default:
            return defaultValue
        }
    }
}
